lib/AST/SearchPathOptions.cpp

```diff
@@ -0,0 +1,96 @@
+#include "llvm/ADT/SmallSet.h"
+
+using namespace swift;
+
+void ModuleSearchPathLookup::addFilesInPathToLookupTable(
+    llvm::vfs::FileSystem *FS, StringRef SearchPath, ModuleSearchPathKind Kind,
+    bool IsSystem, unsigned SearchPathIndex) {
+  std::error_code Error;
+  for (auto Dir = FS->dir_begin(SearchPath, Error);
+       !Error && Dir != llvm::vfs::directory_iterator(); Dir.increment(Error)) {
+    StringRef Filename = llvm::sys::path::filename(Dir->path());
+    LookupTable[Filename].push_back(
+        {SearchPath.str(), Kind, IsSystem, SearchPathIndex});
+  }
+}
+
+void ModuleSearchPathLookup::rebuildLookupTable(const SearchPathOptions *Opts,
+                                                llvm::vfs::FileSystem *FS,
+                                                bool IsOSDarwin) {
+  clearLookupTable();
+
+  for (auto Entry : llvm::enumerate(Opts->getImportSearchPaths())) {
+    addFilesInPathToLookupTable(FS, Entry.value(),
+                                ModuleSearchPathKind::Import,
+                                /*isSystem=*/false, Entry.index());
+  }
+
+  for (auto Entry : llvm::enumerate(Opts->getFrameworkSearchPaths())) {
+    addFilesInPathToLookupTable(FS, Entry.value().Path, ModuleSearchPathKind::Framework,
+                                Entry.value().IsSystem, Entry.index());
+  }
+
+  // Apple platforms have extra implicit framework search paths:
+  // $SDKROOT/System/Library/Frameworks/ and $SDKROOT/Library/Frameworks/.
+  if (IsOSDarwin) {
+    for (auto Entry : llvm::enumerate(Opts->getDarwinImplicitFrameworkSearchPaths())) {
+      addFilesInPathToLookupTable(FS, Entry.value(),
+                                  ModuleSearchPathKind::DarwinImplictFramework,
+                                  /*isSystem=*/true, Entry.index());
+    }
+  }
+
+  for (auto Entry : llvm::enumerate(Opts->getRuntimeLibraryImportPaths())) {
+    addFilesInPathToLookupTable(FS, Entry.value(),
+                                ModuleSearchPathKind::RuntimeLibrary,
+                                /*isSystem=*/true, Entry.index());
+  }
+  State.FileSystem = FS;
+  State.IsOSDarwin = IsOSDarwin;
+  State.Opts = Opts;
+  State.IsPopulated = true;
+}
+
+SmallVector<const ModuleSearchPath *, 4>
+ModuleSearchPathLookup::searchPathsContainingFile(
+    const SearchPathOptions *Opts, llvm::ArrayRef<std::string> Filenames,
+    llvm::vfs::FileSystem *FS, bool IsOSDarwin) {
+  if (!State.IsPopulated || State.FileSystem != FS ||
+      State.IsOSDarwin != IsOSDarwin || State.Opts != Opts) {
+    rebuildLookupTable(Opts, FS, IsOSDarwin);
+  }
+
+  // Gather all search paths that include a file whose name is in Filenames.
+  // To make sure that we don't include search paths twice, keep track of which
+  // search paths have already been added to Result by their kind and Index
+  // in ResultIds.
+  llvm::SmallVector<const ModuleSearchPath *, 4> Result;
+  llvm::SmallSet<std::pair<ModuleSearchPathKind, unsigned>, 4> ResultIds;
```

Comment:
We can still add a search path twice now right?

Reply:
Yes, if it’s specified twice on the command line (and not cleaned up earlier). Is that good or bad or are you just wondering? Search paths could also exist twice as different kinds, e.g. once as framework search path and once as import search path.

Reply:
I was mostly just pointing out `To make sure that we don't include search paths twice` but it is clarified later in that sentence so 🤷

Reply:
Oh, yes. Good catch.